contracts/pcv/uniswap/PCVSwapperUniswap.sol

```diff
@@ -0,0 +1,320 @@
+
+    /// @notice Swap tokenSpent for tokenReceived
+    function swap() external override afterTime whenNotPaused {
+        if (tokenBuyLimit != 0) {
+            require(
+                ERC20(tokenReceived).balanceOf(tokenReceivingAddress) < tokenBuyLimit,
+                "PCVSwapperUniswap: tokenBuyLimit reached."
+            );
+        }
+
+        uint256 amountIn = _getExpectedAmountIn();
+        uint256 amountOut = _getExpectedAmountOut(amountIn);
+        uint256 minimumAcceptableAmountOut = _getMinimumAcceptableAmountOut(amountIn);
+
+        // Check spot price vs oracle price discounted by max slippage
+        // E.g. for a max slippage of 3%, spot price must be >= 97% oraclePrice
+        require(minimumAcceptableAmountOut <= amountOut, "PCVSwapperUniswap: slippage too high.");
+
+        // Reset timer
+        _initTimed();
+
+        // Perform swap
+        ERC20(tokenSpent).safeTransfer(address(pair), amountIn);
+        (uint256 amount0Out, uint256 amount1Out) = pair.token0() == address(tokenSpent)
+            ? (uint256(0), amountOut)
+            : (amountOut, uint256(0));
+        pair.swap(amount0Out, amount1Out, tokenReceivingAddress, new bytes(0));
+
+        // Emit event
+        emit Swap(msg.sender, tokenSpent, tokenReceived, amountIn, amountOut);
+
+        // Incentivize call with FEI rewards
+        _incentivize();
+    }
+
+    // =======================================================================
+    // Internal functions
+    // =======================================================================
+
+    /// @notice incentivize a call with {swapIncentiveAmount} FEI rewards
+    function _incentivize() internal ifMinterSelf {
+        fei().mint(msg.sender, swapIncentiveAmount);
+    }
+
+    /// @notice see external function getNextAmountSpent()
+    function _getExpectedAmountIn() internal view returns (uint256) {
+        uint256 balance = ERC20(tokenSpent).balanceOf(address(this));
+        require(balance != 0, "PCVSwapperUniswap: no tokenSpent left.");
+        return Math.min(maxSpentPerSwap, balance);
+    }
+
+    /// @notice see external function getNextAmountReceived()
+    function _getExpectedAmountOut(uint256 amountIn) internal view returns (uint256) {
+        // Get pair reserves
+        (uint256 _token0, uint256 _token1, ) = pair.getReserves();
+        (uint256 tokenSpentReserves, uint256 tokenReceivedReserves) = pair.token0() == tokenSpent
+            ? (_token0, _token1)
+            : (_token1, _token0);
+
+        // Prepare swap
+        uint256 amountOut = UniswapV2Library.getAmountOut(amountIn, tokenSpentReserves, tokenReceivedReserves);
+
+        return amountOut;
+    }
+
+    /// @notice see external function getNextAmountReceivedThreshold()
+    function _getMinimumAcceptableAmountOut(uint256 amountIn) internal view returns (uint256) {
+        Decimal.D256 memory twap = readOracle();
+        Decimal.D256 memory oracleAmountOut = twap.mul(amountIn);
```

Comment:
be careful about this, most oracles report an USD value, so the "token out" has to be a USD token. Or if oracle reports in ETH value, "token out" has to be in ETH. Easy to misconfigure things around oracles there, we'll need to be very careful in proposal validation scripts when using swappers